python/pyarrow/tests/test_dataset.py

```diff
@@ -451,6 +487,67 @@ def test_dataset(dataset, dataset_reader):
     assert_dataset_fragment_convenience_methods(dataset)
 
 
+def test_dataset_factory_inspect_schema_promotion(promotable_mockfs):
+    mockfs, path1, path2 = promotable_mockfs
+    factory = ds.FileSystemDatasetFactory(
+        mockfs, [path1, path2], ds.ParquetFileFormat()
+    )
+
+    with pytest.raises(
+        pa.ArrowTypeError,
+        match='Unable to merge: Field value has incompatible types: int8 vs uint16',
+    ):
+        factory.inspect()
+
+    schema = factory.inspect(promote_options='permissive')
+    expected_schema = pa.schema([
+        pa.field('value', pa.int32()),
+        pa.field('dictionary', pa.dictionary(pa.int16(), pa.string())),
+    ])
+    assert schema.equals(expected_schema)
+    dataset = factory.finish(schema)
+    table = dataset.to_table()
```

Comment:
Let's validate the table contents for correctness.
```suggestion
    table = dataset.to_table()
    table.validate(full=True)
```